### PR TITLE
[EventHubs] Partition Receiver Options (Track Two)

### PR DESCRIPTION
**Pull request by kinelski**
Included implementation and tests for the class `PartitionReceiverOptions`, which will be used to configure the `PartitionReceiver`. Names are not final and classes are `internal` for the time being.

This implementation is a simple adaptation of the `PartitionReceiver` that was present during previews and follows the [skeleton](https://gist.github.com/jsquire/4ea3368192cfee98dcbb4adfb29a9822) that will be used as a base for the final public API this class will have.

Related: 

- #9323 
- [Partition Receiver Design Proposal](https://gist.github.com/jsquire/4ea3368192cfee98dcbb4adfb29a9822)

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PartitionReceiverOptions.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        private EventHubConnectionOptions _connectionOptions = new EventHubConnectionOptions();
+
+        /// <summary>The set of options to govern retry behavior and try timeouts.</summary>
+        private EventHubsRetryOptions _retryOptions = new EventHubsRetryOptions();
+
+        /// <summary>The amount of time to wait for messages when reading.</summary>
+        private TimeSpan _defaultMaximumReceiveWaitTime = TimeSpan.FromMinutes(1);
+
+        /// <summary>The prefetch count to use for the partition receiver.</summary>
+        private int? _prefetchCount = null;
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PartitionReceiverOptions.cs]: Based on our offline conversation, you're correct in that we should align this to the version from the Event Processor

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PartitionReceiverOptions.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            }
+        }
+
+        /// <summary>
+        ///   The default amount of time to wait for the requested amount of messages when reading; if this
+        ///   period elapses before the requested amount of messages were available or read, then the set of
+        ///   messages that were read will be returned.
+        /// </summary>
+        ///
+        public TimeSpan DefaultMaximumReceiveWaitTime
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PartitionReceiverOptions.cs]: Based on our offline conversation, I'm good with shifting this to nullable, as you recommended.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PartitionReceiverOptions.cs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        ///
+        ///   When an exclusive reader is used, other readers which are non-exclusive or which have a lower owner level will either
+        ///   not be allowed to be created, if they already exist, will encounter an exception during the next attempted operation.
+        /// </summary>
+        ///
+        /// <value>The relative priority to associate with an exclusive reader; for a non-exclusive reader, this value should be <c>null</c>.</value>
+        ///
+        public long? OwnerLevel { get; set; }
+
+        /// <summary>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PartitionReceiverOptions.cs]: Should we fluff this a bit to something closer to the [`EventProcessorOptions`](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessorOptions.cs#L95) text for consistency and to try and explain in a bit more depth?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PartitionReceiver.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+// Licensed under the MIT License.
+
+using System.ComponentModel;
+using System.Threading;
+using Azure.Messaging.EventHubs.Consumer;
+
+namespace Azure.Messaging.EventHubs.Primitives
+{
+    /// <summary>
+    ///   Allows events from a specific partition of an Event Hub, and in the context
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PartitionReceiver.cs]: "Allows  ..."   `reading`?   `receiving`?   ..." events from ..."

**Review by jsquire — COMMENTED** [review]
LGTM.   Waiting to take another pass once you tweak the options to the form that you recommended in conversation.

**Comment by jsquire** [comment]
//cc: @christothes

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PartitionReceiverOptions.cs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        }
+
+        /// <summary>
+        ///   When populated, the owner level indicates that a reading is intended to be performed exclusively for events in the
+        ///   requested partition and for the associated consumer group.  To do so, reading will attempt to assert ownership
+        ///   over the partition; in the case where more than one exclusive reader attempts to assert ownership for the same
+        ///   partition/consumer group pair, the one having a larger <see cref="OwnerLevel"/> value will "win."
+        ///
+        ///   When an exclusive reader is used, other readers which are non-exclusive or which have a lower owner level will either
+        ///   not be allowed to be created, if they already exist, will encounter an exception during the next attempted operation.
```
Comment (ellismg) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PartitionReceiverOptions.cs]: Just wondering.  Do you think we could seealso `EventHubsException.FailureReason.ConsumerDisconnected` here?  Might be nice for folks who have to deal with this to know exactly what the failure mode in the processor will be so they know how to handle it.

**Review by ellismg — APPROVED** [review]

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Consumer/ReadEventOptions.cs [hunk 6]
```diff
@@ -29,6 +29,8 @@ public class ReadEventOptions
         ///
         /// <value>The relative priority to associate with an exclusive reader; for a non-exclusive reader, this value should be <c>null</c>.</value>
         ///
+        /// <seealso cref="EventHubsException.FailureReason.ConsumerDisconnected"/>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Consumer/ReadEventOptions.cs]: We may also want to consider a `<seealso` for the exception type itself above this.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PartitionReceiverOptions.cs [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        /// </summary>
+        ///
+        /// <value>The relative priority to associate with an exclusive reader; for a non-exclusive reader, this value should be <c>null</c>.</value>
+        ///
+        /// <seealso cref="EventHubsException.FailureReason.ConsumerDisconnected"/>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PartitionReceiverOptions.cs]: We probably should expand on this a bit to something like:

```
/// <remarks>
///  A <see cref="EventHubsException">will occur if a <see cref="PartitionReceiver" /> unable to read events from the requested Event Hub partition for the 
///  given consumer group.  In this case the <see cref="EventHubsException.FailureReason"/> will be set to <see cref="EventHubsException.FailureReason.ConsumerDisconnected"/>.
/// </remarks>
```
And then include the `<seealso` for the exception and the one that you have for the reason.

**Review by jsquire — APPROVED** [review]

**Pull request merged**